src/zarr/testing/store.py

```diff
@@ -1,61 +1,121 @@
+from typing import Generic, TypeVar
+
 import pytest
 
 from zarr.abc.store import Store
 from zarr.buffer import Buffer
+from zarr.store.core import _normalize_interval_index
 from zarr.testing.utils import assert_bytes_equal
 
+S = TypeVar("S", bound=Store)
+
+
+class StoreTests(Generic[S]):
+    store_cls: type[S]
 
-class StoreTests:
-    store_cls: type[Store]
+    def set(self, store: S, key: str, value: Buffer) -> None:
+        """
+        Insert key: value pairs into a store without using the store methods.
+        """
+        raise NotImplementedError
+
+    def get(self, store: S, key: str) -> Buffer:
+        """
+        Retrieve values from a store without using the store methods.
+        """
+
+        raise NotImplementedError
 
     @pytest.fixture(scope="function")
     def store(self) -> Store:
         return self.store_cls()
 
-    def test_store_type(self, store: Store) -> None:
+    def test_store_type(self, store: S) -> None:
         assert isinstance(store, Store)
         assert isinstance(store, self.store_cls)
 
-    def test_store_repr(self, store: Store) -> None:
-        assert repr(store)
+    def test_store_repr(self, store: S) -> None:
+        raise NotImplementedError
+
+    def test_store_supports_writes(self, store: S) -> None:
+        raise NotImplementedError
 
-    def test_store_capabilities(self, store: Store) -> None:
-        assert store.supports_writes
-        assert store.supports_partial_writes
-        assert store.supports_listing
+    def test_store_supports_partial_writes(self, store: S) -> None:
+        raise NotImplementedError
+
+    def test_store_supports_listing(self, store: S) -> None:
+        raise NotImplementedError
 
     @pytest.mark.parametrize("key", ["c/0", "foo/c/0.0", "foo/0/0"])
     @pytest.mark.parametrize("data", [b"\x01\x02\x03\x04", b""])
-    async def test_set_get_bytes_roundtrip(self, store: Store, key: str, data: bytes) -> None:
-        await store.set(key, Buffer.from_bytes(data))
-        assert_bytes_equal(await store.get(key), data)
```

Comment:
this test could fail due to bugs in `store.set`, or `store.get`, or both. To disambiguate, I wrote separate tests for `store.set` and `store.get`, that, if both pass, are equivalent to a round-trip test.